In [3]:
import pandas as pd
import plotly.express as px

# data compiled in livepeer_orchestrators.ipynb
# imported here for cleanlinesa
%store -r leaderboard_copy
leaderboard = leaderboard_copy

# note - plotly does not show on github due to javascript blocking. 
# if you clone to your machine and open in vs code / conda the integrated charts should show

In [4]:
# top 10 orchestrators, ordered by staked LPT and their corresponding metrics
# call ratio calculated based on latest 30 rounds
# score is max regional score, on 2/12/2022, to account for the migration on/around 2/14

leaderboard[:10]

transcoders_id  transcoders_totalStake  \
0  0x9c10672cee058fd658103d90872fe431bb6c0afa            2.516184e+06   
1  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9            1.252157e+06   
2  0x525419ff5707190389bfb5c87c375d710f5fcb0e            9.233017e+05   
3  0x4ff088ac5422f994486663ff903b040692797168            7.841820e+05   
4  0x942f0c28fb85ea0b50bfb76a3ecfa99861fa9b4b            7.284690e+05   
5  0xe0a4a877cd0a07da7c08dffebc2546a4713147f2            6.424943e+05   
6  0x4f4758f7167b18e1f5b3c1a7575e3eb584894dbc            5.444993e+05   
7  0x731808ad8b1c3d13e8972db838ada5fc6ae3c2c8            5.117972e+05   
8  0xd84781e1a9b74d71ea76cda8bb9f30893bfd00d1            4.195702e+05   
9  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd            3.410711e+05   

   transcoders_totalVolumeETH  transcoders_totalVolumeUSD  call_ratio  \
0                    8.001220                 1516.763581    0.933333   
1                    0.722895                 1990.434657    1.000000   
2                   21.328402                53037.759233    1.000000   
3                    0.650195                 2203.164261    0.900000   
4                    3.114552                 6065.571979    1.000000   
5                    0.000000                    0.000000    0.900000   
6                    0.628243                 1616.490404    1.000000   
7                    0.000000                    0.000000    0.900000   
8                    1.093463                 2936.040943    0.966667   
9                   10.829665                30211.987059    1.000000   

   delegator_count     score  
0              193         0  
1              263  0.028086  
2              716  0.999986  
3              119         0  
4               14  0.874585  
5              101         0  
6               30         0  
7                2         0  
8              155         0  
9               55  0.987079

Before I even get into the metrics, it's clear that totalStake is not how orchestrators should be ranked, looking at regional scoring. High call_ratios though.

In [7]:
leaderboard.sort_values(by='transcoders_totalVolumeETH', ascending=False)[:10]

transcoders_id  transcoders_totalStake  \
2   0x525419ff5707190389bfb5c87c375d710f5fcb0e            9.233017e+05   
20  0xe3a5793d7c1d2a04a903fa1695b3e3555d6084ca            2.113093e+05   
14  0xbdcbe92befbf36d63ec547bba5842997d77dc841            2.543294e+05   
18  0x9d5611bf0dadddb4441a709141d9229d7f6b3e47            2.145091e+05   
9   0xdac817294c0c87ca4fa1895ef4b972eade99f2fd            3.410711e+05   
22  0x10b21af759129f32c6064adfb85d3ea2a8c0209c            1.324450e+05   
0   0x9c10672cee058fd658103d90872fe431bb6c0afa            2.516184e+06   
10  0xf4e8ef0763bcb2b1af693f5970a00050a6ac7e1b            3.257885e+05   
23  0xe9e284277648fcdb09b8efc1832c73c09b5ecf59            1.162304e+05   
17  0xd0aa1b9d0cd06cafa6af5c1af272be88c38aa831            2.262686e+05   

    transcoders_totalVolumeETH  transcoders_totalVolumeUSD  call_ratio  \
2                    21.328402                53037.759233    1.000000   
20                   13.753365                29565.181273    1.000000   
14                   11.672088                22621.946146    1.000000   
18                   10.954910                25257.496261    1.000000   
9                    10.829665                30211.987059    1.000000   
22                    8.694451                22796.706591    1.000000   
0                     8.001220                 1516.763581    0.933333   
10                    7.886877                12569.912490    0.900000   
23                    5.483522                 8014.533262    1.000000   
17                    4.696978                 9496.728044    0.966667   

    delegator_count     score  
2               716  0.999986  
20               16  0.993759  
14                6  0.990279  
18               30  0.999979  
9                55  0.987079  
22               22  0.999833  
0               193         0  
10               18  0.883412  
23              452  0.864064  
17               35  0.963363

Of the top earners, 7 hold a max regional score >95%. The top orchestrator by stake has earned 8 ETH but, as their score is a 0, I would be curious to see the timeline of when that ETH was earned (i.e. the last time it performed work for the Livepeer network). Using the USD volume as a proxy, that ETH was earned quite a while ago... Further, I'd explore self-delegation - what % of stake is by the orchestrator itself.

In [26]:
board = leaderboard.copy(deep=True)

In [27]:
# calculate percentile rank of each value, for each transcoder

for column in board:
    if column == 'transcoders_id':
        continue
    
    name = f'{column}_pct'
    board[name] = board[column].rank(pct = True)

transcoders_id  transcoders_totalStake  \
0  0x9c10672cee058fd658103d90872fe431bb6c0afa            2.516184e+06   
1  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9            1.252157e+06   
2  0x525419ff5707190389bfb5c87c375d710f5fcb0e            9.233017e+05   
3  0x4ff088ac5422f994486663ff903b040692797168            7.841820e+05   
4  0x942f0c28fb85ea0b50bfb76a3ecfa99861fa9b4b            7.284690e+05   
5  0xe0a4a877cd0a07da7c08dffebc2546a4713147f2            6.424943e+05   
6  0x4f4758f7167b18e1f5b3c1a7575e3eb584894dbc            5.444993e+05   
7  0x731808ad8b1c3d13e8972db838ada5fc6ae3c2c8            5.117972e+05   
8  0xd84781e1a9b74d71ea76cda8bb9f30893bfd00d1            4.195702e+05   
9  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd            3.410711e+05   

   transcoders_totalVolumeETH  transcoders_totalVolumeUSD  call_ratio  \
0                    8.001220                 1516.763581    0.933333   
1                    0.722895                 1990.434657    1.000000   
2                   21.328402                53037.759233    1.000000   
3                    0.650195                 2203.164261    0.900000   
4                    3.114552                 6065.571979    1.000000   
5                    0.000000                    0.000000    0.900000   
6                    0.628243                 1616.490404    1.000000   
7                    0.000000                    0.000000    0.900000   
8                    1.093463                 2936.040943    0.966667   
9                   10.829665                30211.987059    1.000000   

   delegator_count     score  transcoders_totalStake_pct  \
0              193         0                        1.00   
1              263  0.028086                        0.99   
2              716  0.999986                        0.98   
3              119         0                        0.97   
4               14  0.874585                        0.96   
5              101         0                        0.95   
6               30         0                        0.94   
7                2         0                        0.93   
8              155         0                        0.92   
9               55  0.987079                        0.91   

   transcoders_totalVolumeETH_pct  transcoders_totalVolumeUSD_pct  \
0                            0.94                            0.57   
1                            0.65                            0.63   
2                            1.00                            1.00   
3                            0.61                            0.66   
4                            0.87                            0.81   
5                            0.14                            0.14   
6                            0.60                            0.58   
7                            0.14                            0.14   
8                            0.74                            0.70   
9                            0.96                            0.99   

   call_ratio_pct  delegator_count_pct  score_pct  
0           0.770                0.970       0.16  
1           0.915                0.980       0.32  
2           0.915                1.000       0.99  
3           0.730                0.940       0.16  
4           0.915                0.760       0.55  
5           0.730                0.930       0.16  
6           0.915                0.815       0.16  
7           0.730                0.515       0.16  
8           0.805                0.950       0.16  
9           0.915                0.870       0.76

So, looking at the same table, ordered by ETH earned, but with percentile ranks this time. The discrepancy with transcoder 0x9c106... is even clearer here. Though it ranks in the 94% percentile for ETH earned, the USD value of those earnings and score sit at 57% and 16%, respectively.  

In [29]:
board.sort_values(by='transcoders_totalVolumeETH', ascending=False)[:10]

transcoders_id  transcoders_totalStake  \
2   0x525419ff5707190389bfb5c87c375d710f5fcb0e            9.233017e+05   
20  0xe3a5793d7c1d2a04a903fa1695b3e3555d6084ca            2.113093e+05   
14  0xbdcbe92befbf36d63ec547bba5842997d77dc841            2.543294e+05   
18  0x9d5611bf0dadddb4441a709141d9229d7f6b3e47            2.145091e+05   
9   0xdac817294c0c87ca4fa1895ef4b972eade99f2fd            3.410711e+05   
22  0x10b21af759129f32c6064adfb85d3ea2a8c0209c            1.324450e+05   
0   0x9c10672cee058fd658103d90872fe431bb6c0afa            2.516184e+06   
10  0xf4e8ef0763bcb2b1af693f5970a00050a6ac7e1b            3.257885e+05   
23  0xe9e284277648fcdb09b8efc1832c73c09b5ecf59            1.162304e+05   
17  0xd0aa1b9d0cd06cafa6af5c1af272be88c38aa831            2.262686e+05   

    transcoders_totalVolumeETH  transcoders_totalVolumeUSD  call_ratio  \
2                    21.328402                53037.759233    1.000000   
20                   13.753365                29565.181273    1.000000   
14                   11.672088                22621.946146    1.000000   
18                   10.954910                25257.496261    1.000000   
9                    10.829665                30211.987059    1.000000   
22                    8.694451                22796.706591    1.000000   
0                     8.001220                 1516.763581    0.933333   
10                    7.886877                12569.912490    0.900000   
23                    5.483522                 8014.533262    1.000000   
17                    4.696978                 9496.728044    0.966667   

    delegator_count     score  transcoders_totalStake_pct  \
2               716  0.999986                        0.98   
20               16  0.993759                        0.80   
14                6  0.990279                        0.86   
18               30  0.999979                        0.82   
9                55  0.987079                        0.91   
22               22  0.999833                        0.78   
0               193         0                        1.00   
10               18  0.883412                        0.90   
23              452  0.864064                        0.77   
17               35  0.963363                        0.83   

    transcoders_totalVolumeETH_pct  transcoders_totalVolumeUSD_pct  \
2                             1.00                            1.00   
20                            0.99                            0.98   
14                            0.98                            0.95   
18                            0.97                            0.97   
9                             0.96                            0.99   
22                            0.95                            0.96   
0                             0.94                            0.57   
10                            0.93                            0.93   
23                            0.92                            0.88   
17                            0.91                            0.90   

    call_ratio_pct  delegator_count_pct  score_pct  
2            0.915                1.000       0.99  
20           0.915                0.780       0.83  
14           0.915                0.735       0.80  
18           0.915                0.815       0.98  
9            0.915                0.870       0.76  
22           0.915                0.800       0.93  
0            0.770                0.970       0.16  
10           0.730                0.790       0.58  
23           0.915                0.990       0.53  
17           0.805                0.835       0.70

In [46]:
fig_eth = px.bar(
    board[:20],
    x='transcoders_id',
    y='transcoders_totalVolumeETH',
    labels={
        'transcoders_totalVolumeETH':'ETH'
    },
    title='ETH Earned by Top 20 Transcoders (as defined by stake)'
)

fig_eth.show()

This is admittedly going to be arbitrary, but to ultimately determine orchestrator ranking, I am going to use an average of the percentile ranks (excluding totalVolumeUSD). I say arbitrary because a better approach would be to apply weights to each category. However, the weights would ultimately depend on a deeper understanding of what is important as a Livepeer orchestrator (beyond the balance of performance and network security) and what the Livepeer network and community value in orchestrators.  
  
So, straight average it is.

In [81]:
percentiles = board.copy(deep=True)
percentiles.drop(columns=['transcoders_totalStake','transcoders_totalVolumeETH', 'transcoders_totalVolumeUSD', 'call_ratio', 'delegator_count','score','transcoders_totalVolumeUSD_pct'], inplace=True)
percentiles = percentiles.set_index('transcoders_id').T.mean().sort_values(ascending=False)
percentiles


transcoders_id
0x525419ff5707190389bfb5c87c375d710f5fcb0e    0.977
0x9d5611bf0dadddb4441a709141d9229d7f6b3e47    0.900
0xdac817294c0c87ca4fa1895ef4b972eade99f2fd    0.883
0x10b21af759129f32c6064adfb85d3ea2a8c0209c    0.875
0xe3a5793d7c1d2a04a903fa1695b3e3555d6084ca    0.863
                                              ...  
0x90b8b6cf5c923116273fbbee4f146ff83469a1ab    0.202
0x685b3cf5b8fedfddc73f3486983f9432960488ba    0.176
0x750fc601b63554a0d5636f11dbbf8e66a3166889    0.174
0x44c98af5da6a31c422ee5eed18d53986dd9bc01e    0.170
0x3a025488a37fecdb6ca8a4cbd72ba87a919d8bf3    0.164
Length: 100, dtype: float64

In [91]:
top_orchestrators = pd.merge(board, pd.DataFrame(percentiles).rename(columns={0:'avg_pct'}), on='transcoders_id', how='inner')
top10 = top_orchestrators.sort_values(by='avg_pct', ascending=False)[:10]
top10

transcoders_id  transcoders_totalStake  \
2   0x525419ff5707190389bfb5c87c375d710f5fcb0e           923301.742268   
18  0x9d5611bf0dadddb4441a709141d9229d7f6b3e47           214509.084621   
9   0xdac817294c0c87ca4fa1895ef4b972eade99f2fd           341071.081307   
22  0x10b21af759129f32c6064adfb85d3ea2a8c0209c           132444.975869   
20  0xe3a5793d7c1d2a04a903fa1695b3e3555d6084ca           211309.318416   
14  0xbdcbe92befbf36d63ec547bba5842997d77dc841           254329.354468   
12  0x21d1130dc36958db75fbb0e5a9e3e5f5680238ff           280187.427559   
30  0xbe8770603daf200b1fa136ad354ba854928e602b            45875.024295   
28  0x2e3a21ae7cdeb48f57fcad1ce16b258d5502ac05            69690.356908   
23  0xe9e284277648fcdb09b8efc1832c73c09b5ecf59           116230.403060   

    transcoders_totalVolumeETH  transcoders_totalVolumeUSD  call_ratio  \
2                    21.328402                53037.759233         1.0   
18                   10.954910                25257.496261         1.0   
9                    10.829665                30211.987059         1.0   
22                    8.694451                22796.706591         1.0   
20                   13.753365                29565.181273         1.0   
14                   11.672088                22621.946146         1.0   
12                    2.323132                 6764.653814         1.0   
30                    4.336476                15816.668441         1.0   
28                    3.826589                12138.326869         1.0   
23                    5.483522                 8014.533262         1.0   

    delegator_count     score  transcoders_totalStake_pct  \
2               716  0.999986                        0.98   
18               30  0.999979                        0.82   
9                55  0.987079                        0.91   
22               22  0.999833                        0.78   
20               16  0.993759                        0.80   
14                6  0.990279                        0.86   
12               35  0.982702                        0.88   
30               40  0.993716                        0.70   
28                7   0.99969                        0.72   
23              452  0.864064                        0.77   

    transcoders_totalVolumeETH_pct  transcoders_totalVolumeUSD_pct  \
2                             1.00                            1.00   
18                            0.97                            0.97   
9                             0.96                            0.99   
22                            0.95                            0.96   
20                            0.99                            0.98   
14                            0.98                            0.95   
12                            0.82                            0.85   
30                            0.90                            0.94   
28                            0.88                            0.92   
23                            0.92                            0.88   

    call_ratio_pct  delegator_count_pct  score_pct  avg_pct  
2            0.915                1.000       0.99    0.977  
18           0.915                0.815       0.98    0.900  
9            0.915                0.870       0.76    0.883  
22           0.915                0.800       0.93    0.875  
20           0.915                0.780       0.83    0.863  
14           0.915                0.735       0.80    0.858  
12           0.915                0.835       0.75    0.840  
30           0.915                0.860       0.82    0.839  
28           0.915                0.750       0.92    0.837  
23           0.915                0.990       0.53    0.825

So, including a bit of everything gives us a pretty different picture than simply sorting by one metric (reproduced below). 0x525 remains the top orchestrator with an overall percentile rank of .977 - it's no wonder their earnings are by far the highest.

In [90]:
board.sort_values(by='transcoders_totalVolumeETH', ascending=False)[:10]['transcoders_id']

2     0x525419ff5707190389bfb5c87c375d710f5fcb0e
20    0xe3a5793d7c1d2a04a903fa1695b3e3555d6084ca
14    0xbdcbe92befbf36d63ec547bba5842997d77dc841
18    0x9d5611bf0dadddb4441a709141d9229d7f6b3e47
9     0xdac817294c0c87ca4fa1895ef4b972eade99f2fd
22    0x10b21af759129f32c6064adfb85d3ea2a8c0209c
0     0x9c10672cee058fd658103d90872fe431bb6c0afa
10    0xf4e8ef0763bcb2b1af693f5970a00050a6ac7e1b
23    0xe9e284277648fcdb09b8efc1832c73c09b5ecf59
17    0xd0aa1b9d0cd06cafa6af5c1af272be88c38aa831
Name: transcoders_id, dtype: object

In [95]:
fig_score = px.scatter(
    top_orchestrators,
    x='avg_pct',
    y= 'transcoders_totalVolumeETH'
)

fig_score.show()

To no surprise, earnings are concentrated in the orchestrators with the higher overall score.

In [102]:
fig_delegator = px.scatter(
    top_orchestrators,
    x='delegator_count_pct',
    y='transcoders_totalVolumeETH'
)

fig_delegator.show()

Delegator count seems to have minimal bearing on ETH earnings by orchestrator. There are two clusters at .515 and .2 which correspond with the nominal values 2 and 1 (0 would indicate an inactive orchestrator). Removing these from the spread would likely demonstrate an even lower correlation between delegator count and transcoding earnings.

In [107]:
fig_stake = px.scatter(
    top_orchestrators,
    x='transcoders_totalStake_pct',
    y='transcoders_totalVolumeETH'
)

fig_stake.show()

Finally, there is some indication of importance to stake amount and the work on the network (as inferred from fee generation). However, it would appear all nodes in the top quartile have similar earning potential.